test/functional/test-amp-context.js

```diff
@@ -0,0 +1,316 @@
+    expect(context.pageViewId).to.equal('1');
+    expect(context.sentinel).to.equal('1-291921');
+    expect(context.startTime).to.equal('0');
+    expect(context.referrer).to.equal('baz.net');
+  });
+
+  it('should add metadata to window.context using window var.', () => {
+    win.AMP_CONTEXT_DATA = generateAttributes();
+    const context = new AmpContext(win);
+    expect(context).to.be.ok;
+    expect(context.location).to.equal('foo.com');
+    expect(context.canonicalUrl).to.equal('foo.com');
+    expect(context.pageViewId).to.equal('1');
+    expect(context.sentinel).to.equal('1-291921');
+    expect(context.startTime).to.equal('0');
+    expect(context.referrer).to.equal('baz.net');
+  });
+
+  it('should set up only sentinel if no metadata provided.', () => {
+    const sentinel = '123-456';
+    win.AMP_CONTEXT_DATA = sentinel;
+    const context = new AmpContext(win);
+    expect(context).to.be.ok;
+    expect(context.sentinel).to.equal(sentinel);
+  });
+
+  it('should throw error if sentinel invalid', () => {
+    win.name = generateStringifiedAttributes('foobar');
+    const AmpContextSpy = sandbox.spy(AmpContext);
+    try {
+      new AmpContextSpy(win);
+    } catch (err) {
+      // do nothing with it
+    }
+    expect(AmpContextSpy.threw()).to.be.true;
+    expect(AmpContextSpy.exceptions.length).to.equal(1);
+    expect(AmpContextSpy.exceptions[0].message).to.equal(
+        'Incorrect sentinel format');
+  });
+
+  it('should throw error if metadata missing', () => {
+    win.name = generateIncorrectAttributes();
+    const AmpContextSpy = sandbox.spy(AmpContext);
+    try {
+      new AmpContextSpy(win);
+    } catch (err) {
+      // do nothing with it
+    }
+    expect(AmpContextSpy.threw()).to.be.true;
+    expect(AmpContextSpy.exceptions.length).to.equal(1);
+  });
+
+  it('should be able to send an intersection observer request', () => {
+    win.name = generateStringifiedAttributes();
+    const context = new AmpContext(win);
+    const callbackSpy = sandbox.spy();
+    const stopObserving = context.observeIntersection(callbackSpy);
+
+    // window.context should have sent postMessage asking for intersection
+    // observer
+    expect(windowPostMessageSpy.calledOnce).to.be.true;
+    expect(windowPostMessageSpy.calledWith({
+      sentinel: '1-291921',
+      type: MessageType.SEND_INTERSECTIONS,
+    }, '*'));
+
+    // send an intersection message down
+    const messagePayload = {
+      sentinel: '1-291921',
+      type: MessageType.INTERSECTION,
+    };
+    const messageData = 'amp-' + JSON.stringify(messagePayload);
+    const message = {
+      source: context.client_.hostWindow_,
+      data: messageData,
+    };
+    windowMessageHandler(message);
+
+    // window.context should have received intersection observer postMessage
+    // back, and should have called the callback function
+    expect(callbackSpy.calledOnce).to.be.true;
+    expect(callbackSpy.calledWith(messagePayload));
```

Comment:
Side note.

There's a small problem with this that will be fixed once #8469 is merged. The callback for `observeIntersection` in the current window.context API will receive an array of changes, not a change. This should not be a problem for this PR but just letting you know of the current API incompatibility.